chore: cache terraform providers between CI test runs

hugodutka opened the pull request:
Addresses https://github.com/coder/internal/issues/322.

This PR starts caching Terraform providers used by `TestProvision` in `provisioner/terraform/provision_test.go`. The goal is to improve the reliability of this test by cutting down on the number of network calls to external services. It leverages GitHub Actions cache, which [on depot runners is persisted for 14 days by default](https://depot.dev/docs/github-actions/overview#cache-retention-policy).

Other than the aforementioned `TestProvision`, I couldn't find any other tests which depend on external terraform providers.

hugodutka committed:
set up a terraform providers mirror for tests

hugodutka committed:
add cache to CI

spikecurtis commented on .github/actions/setup-test-cache/action.yaml:
```diff
@@ -0,0 +1,55 @@
+        # This prevents PRs from rebuilding the cache on the first day of the month.
+        restore-keys: |
+          ${{ inputs.key-prefix }}-${{ steps.dates.outputs.year-month }}-
+          ${{ github.ref != 'refs/heads/main' && format('{0}-{1}-', inputs.key-prefix, steps.dates.outputs.prev-year-month) || '' }}
```
I don't get why we need to have this special case for PRs. If we are preventing PRs from saving to the cache anyway, why do we need to restrict only PRs from restoring last month's cache if this month is not available?

Also, why are we even bothering to match on month? If today's cache isn't available, doesn't the cache action automatically match the most recent available cache based on the `key-prefix` alone?

hugodutka replied:
This approach lets us reset the cache completely at the start of each month. We upload an updated cache daily, and each update is based on the previous day's cache. Without the monthly cut-off on `main`, any file added to the cache would persist indefinitely. The monthly reset ensures that old or obsolete files do not linger in the cache forever.

spikecurtis replied:
If you implement cleanup in the test code, which I think is a good idea, then we'll only cache what we need daily, and a monthly reset won't be required.

hugodutka replied:
I'm hesitant to get rid of the monthly clean-up for two reasons:

1. If other tests start using the cache and don't clean up properly after themselves, the cache will grow.
2. If a Terraform provider that one of the tests depends on becomes permanently unavailable, we won't be able to detect it without a cache reset.

I don't think matching on the month complicates the code too much, so I'd be for keeping it.

hugodutka commented on provisioner/terraform/provision_test.go:
```diff
@@ -85,6 +91,149 @@ func configure(ctx context.Context, t *testing.T, client proto.DRPCProvisionerCl
+			t.Logf("failed to remove files dir %s: %s", filesDir, err)
+		}
+		if !t.Failed() {
+			return
```
That's to ensure that if `downloadProviders` fails, it doesn't leave an invalid cache directory behind. Since this logic runs in a `defer` statement, it will only delete the directory if the helper function fails - not if the actual test that follows does.

hugodutka commented on provisioner/terraform/provision_test.go:
```diff
@@ -85,6 +91,149 @@ func configure(ctx context.Context, t *testing.T, client proto.DRPCProvisionerCl
+
+	for fileName, file := range templateFiles {
+		filePath := filepath.Join(filesDir, fileName)
+		if _, err := os.Stat(filePath); os.IsNotExist(err) {
```
You're right, this check is redundant. I'll remove it.

spikecurtis commented on provisioner/terraform/provision_test.go:
```diff
@@ -85,6 +91,149 @@ func configure(ctx context.Context, t *testing.T, client proto.DRPCProvisionerCl
+	// Each test gets a unique cache dir based on its name and template files.
+	// This ensures that tests can download providers in parallel and that they
+	// will redownload providers if the template files change.
+	hash := hashTemplateFilesAndTestName(t, t.Name(), templateFiles)
```
Cleanup logic would also be fine here, so we don't blow up people's cache.

hugodutka commented on .github/actions/setup-test-cache/action.yaml:
```diff
@@ -0,0 +1,55 @@
+      # job is running on the main branch.
+      # Without it, PRs would create one-use caches that would linger until
+      # expiration and we'd be charged for them. I evaluated a couple of options
+      # for limiting the cache to the main branch, and forking was the simplest.
```
I considered it but the GitHub Actions syntax makes it unwieldy. I wanted to package all the cache steps into a single action to avoid code duplication across the 5 workflows that need the cache. I tried doing it with `restore` and `save`, but found that [I cannot add a `defer`-like `save` statement in a composite action](https://github.com/orgs/community/discussions/26743). So I'd need to add 3 additional steps to each of the workflows: one to generate the cache keys, one `restore`, one `save`. That's a lot of code duplication - I believe the current approach is cleaner.

spikecurtis replied:
Forking the entire repo seems like a lot of code duplication! It's just better hidden, and means we have yet another fork to manage.

hugodutka replied:
That's a good point. I'll refactor it to use the more verbose restore and save approach.

hugodutka committed:
address feedback: hashing, file exists check, cache clean up

hugodutka committed:
address feedback: use separate save and restore steps, use os-and-arch-specific key prefixes

hugodutka commented:
@spikecurtis I addressed your feedback:

- started using a delimiter when calculating test hashes
- skipped the os.Stat check
- started cleaning up unused test caches
- started using cache key prefixes that include the OS and the Arch
- refactored caching to use separate restore and save steps
